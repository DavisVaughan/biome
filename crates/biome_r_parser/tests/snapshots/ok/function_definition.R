function() 1
function(a) a
