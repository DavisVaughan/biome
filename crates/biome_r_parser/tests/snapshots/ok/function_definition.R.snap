---
source: crates/biome_r_parser/tests/spec_test.rs
expression: snapshot
---
## Input

```R
function() 1
function(a) a
function(a, b) a + b
function(
  a # important!
) a

```


## AST

```
RRoot {
    bom_token: missing (optional),
    expressions: RExpressionList [
        RFunctionDefinition {
            function_token: FUNCTION_KW@0..8 "function" [] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@8..9 "(" [] [],
                items: RParameterList [],
                r_paren_token: R_PAREN@9..10 ")" [] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@10..12 "1" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            function_token: FUNCTION_KW@12..21 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@21..22 "(" [] [],
                items: RParameterList [
                    RParameter {
                        r_identifier: RIdentifier {
                            name_token: IDENT@22..23 "a" [] [],
                        },
                    },
                ],
                r_paren_token: R_PAREN@23..24 ")" [] [],
            },
            body: RIdentifier {
                name_token: IDENT@24..26 "a" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            function_token: FUNCTION_KW@26..35 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@35..36 "(" [] [],
                items: RParameterList [
                    RParameter {
                        r_identifier: RIdentifier {
                            name_token: IDENT@36..37 "a" [] [],
                        },
                    },
                    COMMA@37..38 "," [] [],
                    RParameter {
                        r_identifier: RIdentifier {
                            name_token: IDENT@38..40 "b" [Whitespace(" ")] [],
                        },
                    },
                ],
                r_paren_token: R_PAREN@40..41 ")" [] [],
            },
            body: RBinaryExpression {
                left: RIdentifier {
                    name_token: IDENT@41..43 "a" [Whitespace(" ")] [],
                },
                operator_token_token: PLUS@43..45 "+" [Whitespace(" ")] [],
                right: RIdentifier {
                    name_token: IDENT@45..47 "b" [Whitespace(" ")] [],
                },
            },
        },
        RFunctionDefinition {
            function_token: FUNCTION_KW@47..56 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@56..57 "(" [] [],
                items: RParameterList [
                    RParameter {
                        r_identifier: RIdentifier {
                            name_token: IDENT@57..74 "a" [Newline("\n"), Whitespace("  ")] [Whitespace(" "), Comments("# important!")],
                        },
                    },
                ],
                r_paren_token: R_PAREN@74..76 ")" [Newline("\n")] [],
            },
            body: RIdentifier {
                name_token: IDENT@76..78 "a" [Whitespace(" ")] [],
            },
        },
    ],
    eof_token: EOF@78..79 "" [Newline("\n")] [],
}
```

## CST

```
0: R_ROOT@0..79
  0: (empty)
  1: R_EXPRESSION_LIST@0..78
    0: R_FUNCTION_DEFINITION@0..12
      0: FUNCTION_KW@0..8 "function" [] []
      1: R_PARAMETERS@8..10
        0: L_PAREN@8..9 "(" [] []
        1: R_PARAMETER_LIST@9..9
        2: R_PAREN@9..10 ")" [] []
      2: R_DOUBLE_VALUE@10..12
        0: R_DOUBLE_LITERAL@10..12 "1" [Whitespace(" ")] []
    1: R_FUNCTION_DEFINITION@12..26
      0: FUNCTION_KW@12..21 "function" [Newline("\n")] []
      1: R_PARAMETERS@21..24
        0: L_PAREN@21..22 "(" [] []
        1: R_PARAMETER_LIST@22..23
          0: R_PARAMETER@22..23
            0: R_IDENTIFIER@22..23
              0: IDENT@22..23 "a" [] []
        2: R_PAREN@23..24 ")" [] []
      2: R_IDENTIFIER@24..26
        0: IDENT@24..26 "a" [Whitespace(" ")] []
    2: R_FUNCTION_DEFINITION@26..47
      0: FUNCTION_KW@26..35 "function" [Newline("\n")] []
      1: R_PARAMETERS@35..41
        0: L_PAREN@35..36 "(" [] []
        1: R_PARAMETER_LIST@36..40
          0: R_PARAMETER@36..37
            0: R_IDENTIFIER@36..37
              0: IDENT@36..37 "a" [] []
          1: COMMA@37..38 "," [] []
          2: R_PARAMETER@38..40
            0: R_IDENTIFIER@38..40
              0: IDENT@38..40 "b" [Whitespace(" ")] []
        2: R_PAREN@40..41 ")" [] []
      2: R_BINARY_EXPRESSION@41..47
        0: R_IDENTIFIER@41..43
          0: IDENT@41..43 "a" [Whitespace(" ")] []
        1: PLUS@43..45 "+" [Whitespace(" ")] []
        2: R_IDENTIFIER@45..47
          0: IDENT@45..47 "b" [Whitespace(" ")] []
    3: R_FUNCTION_DEFINITION@47..78
      0: FUNCTION_KW@47..56 "function" [Newline("\n")] []
      1: R_PARAMETERS@56..76
        0: L_PAREN@56..57 "(" [] []
        1: R_PARAMETER_LIST@57..74
          0: R_PARAMETER@57..74
            0: R_IDENTIFIER@57..74
              0: IDENT@57..74 "a" [Newline("\n"), Whitespace("  ")] [Whitespace(" "), Comments("# important!")]
        2: R_PAREN@74..76 ")" [Newline("\n")] []
      2: R_IDENTIFIER@76..78
        0: IDENT@76..78 "a" [Whitespace(" ")] []
  2: EOF@78..79 "" [Newline("\n")] []

```
